interface/ceed.c

```diff
@@ -751,6 +784,66 @@ int CeedReference(Ceed ceed) {
   return CEED_ERROR_SUCCESS;
 }
 
+/**
+  @brief Get a `CeedVector` for scratch work from a `Ceed` context.
+
+  Note: This vector must be restored with @ref CeedRestoreWorkVector().
+
+  @param[in]  ceed `Ceed` context
+  @param[in]  len  Minimum length of work vector
+  @param[out] vec  Address of the variable where `CeedVector` will be stored
+
+  @return An error code: 0 - success, otherwise - failure
+
+  @ref Backend
+**/
+int CeedGetWorkVector(Ceed ceed, CeedSize len, CeedVector *vec) {
+  CeedSize i = 0;
+
+  if (!ceed->work_vectors) CeedCall(CeedWorkVectorsCreate(ceed));
+  // Search for big enough work vector
+  for (i = 0; i < ceed->work_vectors->num_vecs; i++) {
+    if (!ceed->work_vectors->is_in_use[i]) {
+      CeedSize work_len;
+
+      CeedCall(CeedVectorGetLength(ceed->work_vectors->vecs[i], &work_len));
+      if (work_len >= len) break;
+    }
+  }
+
+  // Long enough vector was not found
+  if (i == ceed->work_vectors->num_vecs) {
+    if (ceed->work_vectors->num_vecs == 0) {
+      ceed->work_vectors->num_vecs = 1;
+      CeedCall(CeedCalloc(ceed->work_vectors->num_vecs, &ceed->work_vectors->vecs));
+      CeedCall(CeedCalloc(ceed->work_vectors->num_vecs, &ceed->work_vectors->is_in_use));
+    } else {
+      ceed->work_vectors->num_vecs++;
+      CeedCall(CeedRealloc(ceed->work_vectors->num_vecs, &ceed->work_vectors->vecs));
+      CeedCall(CeedRealloc(ceed->work_vectors->num_vecs, &ceed->work_vectors->is_in_use));
+    }
+    CeedCallBackend(CeedVectorCreate(ceed, len, &ceed->work_vectors->vecs[i]));
```

Comment:
This is going to increment the reference count on `ceed`, which is a reference-counted cycle and thus a memory leak given the current logic in `CeedDestroy`. There are two ways to avoid this. One is for `CeedRestoreWorkVector` to drop its reference to the `Ceed` (putting the object in an invalid state, but we might accept that while it's essentially just a memory allocation, reinstating the `Ceed` when it's checked out) or to count those cyclic references in the `CeedDestroy` logic.

Reply:
Ugh, that's right.

Reply:
ok, should be good now
I'm taking the strategy of leaving un-refcounted vectors in there, bumping the ceed's refcount when checking them out, decreasing the refcount when checking back in